### DISCUSSION TITLE
OIDC client QUARKUS_OIDC_CLIENT_GRANT_OPTIONS__GRANT_NAME_ options

### DISCUSSION THREAD

**Discussion by NMichas** [quarkusio/quarkus#41259] [Q&A] [answered]
For a password grant type, I'm trying to pass `username` and `password` as environment variables. I've tried various naming combinations but everything seems to be ignored, resulting in:
`io.quarkus.oidc.client.OidcClientException: {"error":"invalid_request","error_description":"Missing parameter: username"}`

Reading the relevant documentation, it seems like passing the following should work (but doesn't):
QUARKUS_OIDC_CLIENT_GRANT_OPTIONS__PASSWORD_USERNAME=myuser
![All_configuration_options_-_Quarkus](https://github.com/quarkusio/quarkus/assets/2218296/524d003e-413b-446f-aebe-084fb6154b8d)

Passing it as a parameter with -D works fine, e.g.
`-Dquarkus.oidc-client.grant-options.password.username=myuser`

Could someone please point me to the right direction?

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@NMichas As Far as I understand, due to the slash in the `quarkus.oidc-client` configuration namespace, for it to pick up env properties, one need to declare a property like this `quarkus.oidc-client.grant-options.password.username=` and that should be enough for `QUARKUS_OIDC_CLIENT_GRANT_OPTIONS__PASSWORD_USERNAME=myuser` be effective, give it a try please

CC-ing Roberto, @radcortez as well

**Comment by NMichas** [comment]
Thanks @sberyozkin, you mean to define this property in application properties? I tried adding it to my application.yaml as:
![Notification_Center](https://github.com/quarkusio/quarkus/assets/2218296/c2deaee4-0cb7-4421-bf8b-2c784c5404ea)

And then make sure the env variables are indeed passed (it runs in Kubernetes):
![Notification_Center](https://github.com/quarkusio/quarkus/assets/2218296/ce0b532e-ea35-494b-b822-14df093c00b3)

Now I get:
`io.quarkus.oidc.client.OidcClientException: {"error":"invalid_grant","error_description":"Invalid user credentials"}`

So, I guess Quarkus is using the dummy values I provided in application.yaml and the environment variables are still ignored.

**Comment by dmlloyd** [comment, marked as answer]
The environment variable in this case is incorrect; remove the double `__` and replace it with a single `_` and it should work.

Apart from a few specific circumstances, env vars cannot _establish_ configuration, they can only _override_ configuration.

**Comment by NMichas** [comment]
Indeed, that worked, thanks a lot!

**Reply by sberyozkin** [reply]
Thanks @dmlloyd, @NMichas
